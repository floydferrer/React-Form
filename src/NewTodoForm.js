import React, { useState } from "react";

const NewTodoForm = ({ addTodo }) => {
    const initialState = {
        name: ''
    }

    const [formData, setFormData] = useState(initialState);
    const [isInvalid, setIsInvalid] = useState(true);
    const [isTouched, setIsTouched] = useState(false);
    const handleChange = e => {
        setIsTouched(true);
        const {name, value} = e.target;
        if(value === '') {
            setIsInvalid(true);
        } else {
            setIsInvalid(false);
        }

        setFormData(data => ({
            ...data,
            [name]: value
        }))
    }

    const handleSubmit = (e) => {
        e.preventDefault();
        const { name } = formData;
        if(!isInvalid) {
            alert(`New task added!`);
            addTodo({...formData})
            setFormData(initialState);
        }
    }

    return (
        <form onSubmit={handleSubmit}>
            <h3>Create Todo</h3>
            <label htmlFor="name">Name</label>
            <input
                type="text"
                placeholder="name"
                name="name"
                id="name"
                value={formData.name}
                onChange={handleChange}
            />
            {isInvalid && isTouched && <span style={{color: 'red'}}>Color cannot be blank!</span>}
            
            <button role="button">Add Todo</button>
        </form>
    )
}

export default NewTodoForm;