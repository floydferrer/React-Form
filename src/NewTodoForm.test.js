import { render, fireEvent } from "@testing-library/react";
import NewTodoForm from "./NewTodoForm";

it("renders without crashing", function() {
    render(<NewTodoForm />)
});

it("matches snapshot", function() {
    const { asFragment } = render(<NewTodoForm />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("works when you submit form", function() {
    const addTodoMock = jest.fn();
    const { getByLabelText, getByRole } = render(<NewTodoForm addTodo={addTodoMock}/>)
    const taskInput = getByLabelText("Name");
    fireEvent.change(taskInput, { target: { value: 'test task' }});
    const addBtn = getByRole('button');
    fireEvent.click(addBtn);

    expect(addTodoMock).toHaveBeenCalled();
  });