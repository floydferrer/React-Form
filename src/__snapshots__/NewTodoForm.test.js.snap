// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form>
    <h3>
      Create Todo
    </h3>
    <label
      for="name"
    >
      Name
    </label>
    <input
      id="name"
      name="name"
      placeholder="name"
      type="text"
      value=""
    />
    <button
      role="button"
    >
      Add Todo
    </button>
  </form>
</DocumentFragment>
`;
